lib/cretonne/src/ir/frontend.rs

```diff
@@ -0,0 +1,764 @@
+//! A frontend for building Cretonne IL from other languages.
+//!
+//! # Example
+//!
+//! Here is a pseudo-program we want to transform in Cretonne IL:
+//!
+//! ```cton
+//! function(x) {
+//! x, y, z : I32
+//! block0:
+//!    y = 2;
+//!    z = x + y;
+//!    jump block1
+//! block1:
+//!    z = z + y;
+//!    brnz y, block2;
+//!    z = z - x;
+//!    return y
+//! block2:
+//!    y = y - x
+//!    jump block1
+//! }
+//! ```
+//!
+//! Here is how you build the corresponding Cretonne IL function using `ILBuilder`:
+//!
+//! ```rust
+//! use cretonne::entity_ref::EntityRef;
+//! use cretonne::ir::{FunctionName, Function, Signature, ArgumentType, InstBuilder};
+//! use cretonne::ir::types::*;
+//! use cretonne::ir::frontend::{ILBuilder, FunctionBuilder};
+//! use cretonne::verifier::verify_function;
+//!
+//! use std::u32;
+//!
+//! // An opaque reference to variable.
+//! #[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+//! pub struct Variable(u32);
+//! impl EntityRef for Variable {
+//!     fn new(index: usize) -> Self {
+//!         assert!(index < (u32::MAX as usize));
+//!         Variable(index as u32)
```

Comment:
I can do that but you should go check out `entity_ref.rs`, I copied the code from there and those remarks should be addressed more generally to the rest of Cretonne.